File: webtransport-py/broadsock.py
from typing import List, Dict
from app_logs import getLogger
from app_models import Client, ReliableConnection, FastUnreliableConnection, PLAYER_TYPE_FAMILY, PLAYER_TYPE_SURVIVOR, Rooms, ClientGameMessages, GameServerMessages, get_uid_from_msg

Log = getLogger(__name__)


uid_sequence = 0
clients: List[Client] = []
disconnected_clients: List[Client] = []
reliable_connection = ReliableConnection(clients)
fast_unreliable_connection = FastUnreliableConnection(reliable_connection)
rooms = Rooms()

def get_next_uid_sequence() -> int:
    global uid_sequence
    return uid_sequence + 1

def has_connected_clients() -> bool:
    return len(clients) > 0

def get_client_by_ws(ws) -> Client:
    global clients
    # Log.info(f'get_client_by_ws clients: {len(clients)}')
    for c in clients:
        if c.reliableWS is ws:
            return c

def get_client_by_wt(wt) -> Client:
    global clients
    # Log.info(f'get_client_by_ws clients: {len(clients)}')
    for c in clients:
        if c.unreliableFastWT is wt:
            return c

"""
Client connect/disconnect
"""
def handle_client_connected(websocket) -> Client:
    global uid_sequence

    if websocket is not None:
        client = get_client_by_ws(websocket)

    uid_sequence += 1

    if client is None:
        client = Client(uid_sequence, None, websocket, None)
        clients.append(client)

    if client.reliableWS is None and websocket is not None:
        client.reliableWS = websocket

    Log.info(f'client connected: {client}, clients = {len(clients)}')

    return client

async def handle_client_disconnected(websocket):
    Log.debug(f'clients = {", ".join(map(str, clients))}: disconnecting..., WS: {id(websocket)}')

    if len(clients) == 0:
        return

    client = get_client_by_ws(websocket)

    if client.uid is not None:
        to_game_server(GameServerMessages.DISCONNECT, client)

    client.disconnect()
    disconnected_clients.append(client)
    clients.remove(client)

    # rooms.remove_room_by_player(client.uid)
    rooms.remove_player(client)

    await reliable_connection.send_message_all(f'{rooms}')
    await reliable_connection.send_message_all(f'-1.ONLINE.{len(clients)}')

    Log.info(f'client disconected: {client}, clients = {len(clients)}')

    if len(clients) == 0:
        Log.info(f'clients = {len(clients)}: Need stop game server')


"""
Set client/server connection
"""
async def set_game_client_communication_websocket(websocket) -> Client:
    client = handle_client_connected(websocket)
    await reliable_connection.send_message_all(f'{rooms}')
    return client

def set_game_client_communication_web_transport(c_uid: int, web_transport) -> Client:
    for client in clients:
        if client.uid == c_uid:
            client.unreliableFastWT = web_transport
            return client

async def send_usernames():
    res = ClientGameMessages.USERNAMES
    for client in clients:
        res += f'{client.uid}.{len(client.username)}#{client.username}'
    await reliable_connection.send_message_all(f'{res}')

"""
Send messages to server and client
"""
async def to_server(msg, client: Client):
    Log.debug(f'TO-SERVER: {msg}, client: {client}')

    global rooms
    send_to_server = True

    if GameServerMessages.CONNECT_ME in msg:
        uid = get_uid_from_msg(msg)
        disconnected = None
        if uid:
            for c in disconnected_clients:
                if c.uid == uid:
                    disconnected = c

        if disconnected:
            disconnected_clients.remove(disconnected)
            if disconnected.can_connect():
                client.uid = uid
                client.username = f'user-{uid}'

        await reliable_connection.send_msg_to(client, f'{client.uid}.CONNECT_SELF.{client.username}')
        await reliable_connection.send_message_others(f'{client.uid}.CONNECT_OTHER', client.uid)
        await reliable_connection.send_message_all(f'-1.ONLINE.{len(clients)}')
        await send_usernames()
        return

    if ClientGameMessages.GET_USERNAMES in msg:
        await send_usernames()
        return

    if ClientGameMessages.ROOMS_GET in msg:
        send_to_server = False
    # elif 'NOT_GS_CREATE_ROOM' in msg:
    #     rooms.add(Room(f'Room {rooms.size()}'))
    #     send_to_server = False
    elif ClientGameMessages.WS_PONG in msg:
        client.calc_ws_latency()
        return
    elif ClientGameMessages.JOIN_ROOM in msg:
        parts = msg.split('.')
        room_name = parts[1]
        player_type = PLAYER_TYPE_SURVIVOR
        if parts[2] == 'family':
            player_type = PLAYER_TYPE_FAMILY
        rooms.add_player(room_name, player_type, client)
        send_to_server = False
    elif ClientGameMessages.PLAYER_READY in msg:
        parts = msg.split('.')
        await rooms.player_ready(parts[1], client)
        send_to_server = False
    elif ClientGameMessages.SET_PLAYER_USERNAME in msg:
        client.set_username(msg[27:])
        await send_usernames()
        return

    if send_to_server:
        to_game_server(msg, client)
    else:
        await reliable_connection.send_message_all(f'{rooms}')

def to_game_server(msg, client: Client):
    msg = f'{client.uid}.{msg}'

    room = rooms.get_room_by_client_uid(client.uid)
    if room is not None and room.game_server is not None:
        room.game_server.write(msg)

File: game/src/main/java/ovh/look/game/websocket/BroadSock.java
package ovh.look.game.websocket;

import org.springframework.stereotype.Service;
import org.springframework.web.reactive.socket.WebSocketSession;
import ovh.look.game.models.*;
import ovh.look.game.rooms.Room;
import ovh.look.game.rooms.Rooms;
import ovh.look.game.server.GameServerMessages;
import java.util.ArrayList;
import java.util.List;
import java.util.logging.Logger;

@Service
public class BroadSock {

    private static final Logger Log = Logger.getLogger(BroadSock.class.getName());

    private static int uidSequence = 0;
    private static List<Client> clients = new ArrayList<>();
    private static List<Client> disconnectedClients = new ArrayList<>();
    private static ReliableConnection reliableConnection = new ReliableConnection(clients);
    
    private final Rooms rooms;

    public BroadSock(Rooms rooms) {
        this.rooms = rooms;
    }

    public static int getNextUidSequence() {
        return uidSequence + 1;
    }

    public static boolean hasConnectedClients() {
        return !clients.isEmpty();
    }

    public static Client getClientByWs(WebSocketSession ws) {
        for (Client client : clients) {
            if (client.getReliableWS() != null
                    && client.getReliableWS().getId().equals(ws.getId())) {
                return client;
            }
        }
        return null;
    }

    public static Client getClientByUid(int uid) {
        for (Client client : clients) {
            if (client.getUid() == uid) {
                return client;
            }
        }
        return null;
    }

    public static void addClient(Client client) {
        clients.add(client);
        Log.info("Client " + client.getUid() + " added.");
    }

    public static void removeClient(Client client) {
        clients.remove(client);
        Log.info("Client " + client.getUid() + " removed.");
    }

    /*
     * Client connect/disconnect
     */
    public Client handleClientConnected(WebSocketSession session, int uid) {
        Client client = getClientByWs(session);
        if (client == null) {
            if (uid <= 0) {
                uidSequence += 1;
                client = new Client(uidSequence, null, session, null);
                clients.add(client);
            } else {
                client = getClientByUid(uid);
                if (client == null) {
                    uidSequence += 1;
                    client = new Client(uidSequence, null, session, null);
                    clients.add(client);
                } else {
                    client.setReliableWS(session);
                }
            }
        }

        client.connect();

        Log.info(String.format("client connected: %s, clients = %d", client, clients.size()));

        return client;
    }

    public void handleClientDisconnected(WebSocketSession session) {
        // Log.debug(f'clients = {", ".join(map(str, clients))}: disconnecting..., WS: {id(websocket)}')
        if (clients.isEmpty()) return;

        Client client = getClientByWs(session);

        if (client.getUid() > 0) {
            toGameServer(GameServerMessages.DISCONNECT.getValue(), client);
        }

        client.disconnect();
        disconnectedClients.add(client);
        clients.remove(client);

        rooms.removePlayer(client);

        reliableConnection.sendMessageAll(rooms.toString());
        reliableConnection.sendMessageAll("-1.ONLINE." + clients.size());

        Log.info(String.format("client disconected: %d, clients = %d", client.getUid(), clients.size()));

        if (clients.isEmpty()) {
            Log.info(String.format("clients = %d: Need stop game server", clients.size()));
        }
    }

    /*
     * Set client/server connection
     */
    public Client setGameClientCommunicationWebSocket(WebSocketSession websocket, int uid) {
        Client client = handleClientConnected(websocket, uid);
        reliableConnection.sendMessageAll(rooms.toString());
        return client;
    }

    public static Client setGameClientCommunicationWebTransport(int cUid, Object webTransport) {
        for (Client client : clients) {
            if (client.getUid() == cUid) {
                client.setUnreliableFastWT(webTransport);
                return client;
            }
        }
        return null;
    }

    public static void sendUsernames() {
        StringBuilder res = new StringBuilder(ClientGameMessages.USERNAMES.getValue());
        for (Client client : clients) {
            res.append(client.getUid()).append(".").append(client.getUsername().length()).append("#").append(client.getUsername());
        }
        reliableConnection.sendMessageAll(res.toString());
    }

    /*
     * Send messages to server and client
     */
    public void toServer(String msg, WebSocketSession session) {
        Client client = getClientByWs(session);

        Log.info("TO-SERVER: " + msg + ", client: " + (client != null ? client.getUsername() : null));

        boolean sendToServer = true;

        if (client != null && msg.contains(GameServerMessages.CONNECT_ME.getValue())) {
            String[] parts = msg.split("\\.");
            int uid = parts.length > 1 ? Integer.parseInt(parts[1]) : 0;
            Client disconnected = null;

            if (uid > 0) {
                for (var c: disconnectedClients) {
                    if (c.getUid() == uid) {
                        disconnected = c;
                        break;
                    }
                }

                if (disconnected != null) {
                    disconnectedClients.remove(disconnected);
                    if (disconnected.canConnect()) {
                        client.setUid(uid);
                        client.setUsername("user-" + uid);
                    }
                }
            }

            reliableConnection.sendMsgTo(client, client.getUid() + ".CONNECT_SELF." + client.getUsername());
            reliableConnection.sendMessageOthers(client.getUid() + ".CONNECT_OTHER", client.getUid());
            reliableConnection.sendMessageAll("-1.ONLINE." + clients.size());
            sendUsernames();
            return;
        }

        if (msg.contains(ClientGameMessages.GET_USERNAMES.getValue())) {
            sendUsernames();
            return;
        }

        if (msg.contains(ClientGameMessages.ROOMS_GET.getValue())) {
            sendToServer = false;
        } else if (msg.contains(ClientGameMessages.WS_PONG.getValue())) {
            client.calcWsLatency();
            return;
        } else if (msg.contains(ClientGameMessages.JOIN_ROOM.getValue())) {
            String[] parts = msg.split("\\.");
            String roomName = parts[1];
            // int playerType = parts[2].equals("family") ? PlayerType.FAMILY.ordinal() : PlayerType.SURVIVOR.ordinal();
            PlayerType playerType = parts[2].equals("family") ? PlayerType.FAMILY : PlayerType.SURVIVOR;
            rooms.addPlayer(roomName, playerType, client);
            sendToServer = false;
        } else if (msg.contains(ClientGameMessages.PLAYER_READY.getValue())) {
            String[] parts = msg.split("\\.");
            rooms.playerReady(parts[1], client);
            sendToServer = false;
        } else if (msg.contains(ClientGameMessages.SET_PLAYER_USERNAME.getValue())) {
            client.setUsername(msg.substring(27));
            sendUsernames();
            return;
        }

        if (sendToServer) {
            toGameServer(msg, client);
        } else {
            reliableConnection.sendMessageAll(rooms.toString());
        }
    }

    private void toGameServer(String msg, Client client) {
        msg = client.getUid() + "." + msg;
        Room room = rooms.getRoomByClientUid(client.getUid());
        if (room != null && room.getGameServer() != null) {
            room.getGameServer().write(msg);
        }
    }
}
